File: backoff.c
#include "backoff.h"
#include <stdio.h>
#include <unistd.h>

inline void backoff_init(struct backoff *b, useconds_t min, useconds_t max) {
  b->max = max;
  b->min = min;
  backoff_clear(b);
} /* backoff_init */

inline void backoff(struct backoff *b) {
  //printf("Sleeping %f seconds\n", b->time / 1000000.0);
  usleep(b->time);

  /* Exponential backoff */
  b->time <<= 1;

  /* Cap at 'max' time sleep */
  if (b->time > b->max) {
    b->time = b->max;
  }
} /* backoff_sleep */

inline void backoff_clear(struct backoff *b) {
  b->time = b->min; /* 1000 microseconds == 1ms */
} /* backoff_clear */

File: go/src/backoff/backoff.go
package backoff

import (
  "time"
)

type Backoff struct {
  min time.Duration
  max time.Duration
  cur time.Duration
}

func NewBackoff(min time.Duration, max time.Duration) (*Backoff) {
  return &Backoff{min, max, min}
}

func (b *Backoff) Wait() {
  time.Sleep(b.cur)

  b.cur *= 2
  if (b.cur > b.max) {
    b.cur = b.max
  }
}

func (b *Backoff) Reset() {
  b.cur = b.min
}
